File: Flutter/IMSDK/im-flutter-plugin/tencent_im_sdk_plugin/android/src/main/java/com/qq/qcloud/tencent_im_sdk_plugin/manager/ConversationManager.java
package com.qq.qcloud.tencent_im_sdk_plugin.manager;

import com.qq.qcloud.tencent_im_sdk_plugin.util.CommonUtil;
import com.tencent.imsdk.v2.V2TIMCallback;
import com.tencent.imsdk.v2.V2TIMConversation;
import com.tencent.imsdk.v2.V2TIMConversationListFilter;
import com.tencent.imsdk.v2.V2TIMConversationListener;
import com.tencent.imsdk.v2.V2TIMConversationOperationResult;
import com.tencent.imsdk.v2.V2TIMConversationResult;
import com.tencent.imsdk.v2.V2TIMManager;
import com.tencent.imsdk.v2.V2TIMValueCallback;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class ConversationManager {
    private static MethodChannel channel;
    private static  HashMap<String, V2TIMConversationListener> conversationListenerList= new HashMap();
    public ConversationManager(MethodChannel _channel){
        ConversationManager.channel = _channel;
    }

    public  void setConversationListener(MethodCall methodCall, final MethodChannel.Result result){
        final String listenerUuid = methodCall.argument("listenerUuid");
        V2TIMManager.getConversationManager().setConversationListener(new V2TIMConversationListener() {
            @Override
            public void onSyncServerStart() {
                makeConversationListenerEventData("onSyncServerStart",null, listenerUuid);
            }

            @Override
            public void onSyncServerFinish() {
                makeConversationListenerEventData("onSyncServerFinish",null, listenerUuid);
            }

            @Override
            public void onSyncServerFailed() {
                makeConversationListenerEventData("onSyncServerFailed",null, listenerUuid);
            }

            @Override
            public void onNewConversation(List<V2TIMConversation> conversationList) {
                LinkedList<HashMap<String,Object>> list = new LinkedList<HashMap<String,Object>>();
                for(int i = 0;i<conversationList.size();i++){
                    list.add(CommonUtil.convertV2TIMConversationToMap(conversationList.get(i)));
                }
                makeConversationListenerEventData("onNewConversation",list, listenerUuid);
            }

            @Override
            public void onConversationChanged(List<V2TIMConversation> conversationList) {
                LinkedList<HashMap<String,Object>> list = new LinkedList<HashMap<String,Object>>();
                for(int i = 0;i<conversationList.size();i++){
                    list.add(CommonUtil.convertV2TIMConversationToMap(conversationList.get(i)));
                }
                makeConversationListenerEventData("onConversationChanged",list, listenerUuid);
            }
            @Override
            public void onTotalUnreadMessageCountChanged(long totalUnreadCount) {
                makeConversationListenerEventData("onTotalUnreadMessageCountChanged",totalUnreadCount, listenerUuid);
            }
            @Override
            public void onConversationGroupCreated(String groupName, List<V2TIMConversation> conversationList) {
                HashMap<String,Object> data = new HashMap<>();
                data.put("groupName",groupName);
                LinkedList<HashMap<String,Object>> list = new LinkedList<HashMap<String,Object>>();
                for(int i = 0;i<conversationList.size();i++){
                    list.add(CommonUtil.convertV2TIMConversationToMap(conversationList.get(i)));
                }
                data.put("conversationList",list);
                makeConversationListenerEventData("onConversationGroupCreated",data, listenerUuid);
            }

            @Override
            public void onConversationGroupDeleted(String groupName) {
                makeConversationListenerEventData("onConversationGroupDeleted",groupName, listenerUuid);
            }

            @Override
            public void onConversationGroupNameChanged(String oldName, String newName) {
                HashMap<String,Object> data = new HashMap<>();
                data.put("oldName",oldName);
                data.put("newName",newName);
                makeConversationListenerEventData("onConversationGroupNameChanged",data, listenerUuid);
            }

            @Override
            public void onConversationsAddedToGroup(String groupName, List<V2TIMConversation> conversationList) {
                HashMap<String,Object> data = new HashMap<>();
                data.put("groupName",groupName);
                LinkedList<HashMap<String,Object>> list = new LinkedList<HashMap<String,Object>>();
                for(int i = 0;i<conversationList.size();i++){
                    list.add(CommonUtil.convertV2TIMConversationToMap(conversationList.get(i)));
                }
                data.put("conversationList",list);
                makeConversationListenerEventData("onConversationsAddedToGroup",data, listenerUuid);
            }

            @Override
            public void onConversationsDeletedFromGroup(String groupName, List<V2TIMConversation> conversationList) {
                HashMap<String,Object> data = new HashMap<>();
                data.put("groupName",groupName);
                LinkedList<HashMap<String,Object>> list = new LinkedList<HashMap<String,Object>>();
                for(int i = 0;i<conversationList.size();i++){
                    list.add(CommonUtil.convertV2TIMConversationToMap(conversationList.get(i)));
                }
                data.put("conversationList",list);
                makeConversationListenerEventData("onConversationsDeletedFromGroup",data, listenerUuid);
            }
        });
        result.success("add conversation listener success");
    }
    public void removeConversationListener(MethodCall call,final MethodChannel.Result result){
        final String listenerUuid = CommonUtil.getParam(call,result,"listenerUuid");
        if (listenerUuid != "") {
            final V2TIMConversationListener listener = conversationListenerList.get(listenerUuid);
            V2TIMManager.getConversationManager().removeConversationListener(listener);
            conversationListenerList.remove(listenerUuid);
            result.success("removeConversationListener is done");
        } else {
            for (V2TIMConversationListener listener : conversationListenerList.values()) {
                V2TIMManager.getConversationManager().removeConversationListener(listener);
            }
            conversationListenerList.clear();
            result.success("all conversation listener message is removed");
        }
    }

    public void  addConversationListener(MethodCall call,final MethodChannel.Result result){
        final String listenerUuid = CommonUtil.getParam(call,result,"listenerUuid");
        final V2TIMConversationListener conversationLister = new V2TIMConversationListener() {
            @Override
            public void onSyncServerStart() {
                makeConversationListenerEventData("onSyncServerStart",null, listenerUuid);
            }

            @Override
            public void onSyncServerFinish() {
                makeConversationListenerEventData("onSyncServerFinish",null, listenerUuid);
            }

            @Override
            public void onSyncServerFailed() {
                makeConversationListenerEventData("onSyncServerFailed",null, listenerUuid);
            }

            @Override
            public void onNewConversation(List<V2TIMConversation> conversationList) {
                LinkedList<HashMap<String,Object>> list = new LinkedList<HashMap<String,Object>>();
                for(int i = 0;i<conversationList.size();i++){
                    list.add(CommonUtil.convertV2TIMConversationToMap(conversationList.get(i)));
                }
                makeConversationListenerEventData("onNewConversation",list, listenerUuid);
            }

            @Override
            public void onConversationChanged(List<V2TIMConversation> conversationList) {
                LinkedList<HashMap<String,Object>> list = new LinkedList<HashMap<String,Object>>();
                for(int i = 0;i<conversationList.size();i++){
                    list.add(CommonUtil.convertV2TIMConversationToMap(conversationList.get(i)));
                }
                makeConversationListenerEventData("onConversationChanged",list, listenerUuid);
            }
            @Override
            public void onTotalUnreadMessageCountChanged(long totalUnreadCount) {
                makeConversationListenerEventData("onTotalUnreadMessageCountChanged",totalUnreadCount, listenerUuid);
            }

            @Override
            public void onConversationGroupCreated(String groupName, List<V2TIMConversation> conversationList) {
                HashMap<String,Object> data = new HashMap<>();
                data.put("groupName",groupName);
                LinkedList<HashMap<String,Object>> list = new LinkedList<HashMap<String,Object>>();
                for(int i = 0;i<conversationList.size();i++){
                    list.add(CommonUtil.convertV2TIMConversationToMap(conversationList.get(i)));
                }
                data.put("conversationList",list);
                makeConversationListenerEventData("onConversationGroupCreated",data, listenerUuid);
            }

            @Override
            public void onConversationGroupDeleted(String groupName) {
                makeConversationListenerEventData("onConversationGroupDeleted",groupName, listenerUuid);
            }

            @Override
            public void onConversationGroupNameChanged(String oldName, String newName) {
                HashMap<String,Object> data = new HashMap<>();
                data.put("oldName",oldName);
                data.put("newName",newName);
                makeConversationListenerEventData("onConversationGroupNameChanged",data, listenerUuid);
            }

            @Override
            public void onConversationsAddedToGroup(String groupName, List<V2TIMConversation> conversationList) {
                HashMap<String,Object> data = new HashMap<>();
                data.put("groupName",groupName);
                LinkedList<HashMap<String,Object>> list = new LinkedList<HashMap<String,Object>>();
                for(int i = 0;i<conversationList.size();i++){
                    list.add(CommonUtil.convertV2TIMConversationToMap(conversationList.get(i)));
                }
                data.put("conversationList",list);
                makeConversationListenerEventData("onConversationsAddedToGroup",data, listenerUuid);
            }

            @Override
            public void onConversationsDeletedFromGroup(String groupName, List<V2TIMConversation> conversationList) {
                HashMap<String,Object> data = new HashMap<>();
                data.put("groupName",groupName);
                LinkedList<HashMap<String,Object>> list = new LinkedList<HashMap<String,Object>>();
                for(int i = 0;i<conversationList.size();i++){
                    list.add(CommonUtil.convertV2TIMConversationToMap(conversationList.get(i)));
                }
                data.put("conversationList",list);
                makeConversationListenerEventData("onConversationsDeletedFromGroup",data, listenerUuid);
            }
        };
        conversationListenerList.put(listenerUuid, conversationLister);
        V2TIMManager.getConversationManager().addConversationListener(conversationLister);
        result.success("addConversationListener success");
    }


    public  void  getConversation(MethodCall methodCall,final  MethodChannel.Result result){
        String conversationID = CommonUtil.getParam(methodCall,result,"conversationID");
        V2TIMManager.getConversationManager().getConversation(conversationID, new V2TIMValueCallback<V2TIMConversation>() {
            @Override
            public void onError(int i, String s) {
                CommonUtil.returnError(result,i,s);
            }

            @Override
            public void onSuccess(V2TIMConversation v2TIMConversation) {
                CommonUtil.returnSuccess(result,CommonUtil.convertV2TIMConversationToMap(v2TIMConversation));
            }
        });
    }
    private <T> void  makeConversationListenerEventData(String type,T data, String listenerUuid){
        CommonUtil.emitEvent(ConversationManager.channel,"conversationListener",type,data, listenerUuid);
    }
    public void getConversationList(MethodCall methodCall, final MethodChannel.Result result){
        String nextSeq = CommonUtil.getParam(methodCall,result,"nextSeq");
        int count = CommonUtil.getParam(methodCall,result,"count");
        V2TIMManager.getConversationManager().getConversationList(Long.parseLong(nextSeq), count, new V2TIMValueCallback<V2TIMConversationResult>() {
            @Override
            public void onError(int i, String s) {
                CommonUtil.returnError(result,i,s);
            }

            @Override
            public void onSuccess(V2TIMConversationResult v2TIMConversationResult) {
                CommonUtil.returnSuccess(result,CommonUtil.convertV2TIMConversationResultToMap(v2TIMConversationResult));
            }
        });
    }
    public void getConversationListByConversaionIds(MethodCall methodCall, final MethodChannel.Result result){
      List<String>  conversationIDList = CommonUtil.getParam(methodCall,result,"conversationIDList");
      V2TIMManager.getConversationManager().getConversationList(conversationIDList, new V2TIMValueCallback<List<V2TIMConversation>>() {
          @Override
          public void onSuccess(List<V2TIMConversation> v2TIMConversations) {
              LinkedList<HashMap<String,Object>> list = new LinkedList<HashMap<String, Object>>();
              for(int i = 0;i<v2TIMConversations.size();i++){
                  list.add(CommonUtil.convertV2TIMConversationToMap(v2TIMConversations.get(i)));
              }
              CommonUtil.returnSuccess(result,list);
          }

          @Override
          public void onError(int i, String s) {
                CommonUtil.returnError(result,i,s);
          }
      });
    }
    public void pinConversation(MethodCall methodCall, final MethodChannel.Result result){
        String conversationID = CommonUtil.getParam(methodCall,result,"conversationID");
       boolean isPinned = CommonUtil.getParam(methodCall,result,"isPinned");
       V2TIMManager.getConversationManager().pinConversation(conversationID, isPinned, new V2TIMCallback() {
           @Override
           public void onSuccess() {
               CommonUtil.returnSuccess(result,null);
           }

           @Override
           public void onError(int i, String s) {
                CommonUtil.returnError(result,i,s);
           }
       });
    }
    public void getTotalUnreadMessageCount(MethodCall methodCall, final MethodChannel.Result result){
        V2TIMManager.getConversationManager().getTotalUnreadMessageCount(new V2TIMValueCallback<Long>() {
            @Override
            public void onSuccess(Long aLong) {
                CommonUtil.returnSuccess(result,aLong);
            }

            @Override
            public void onError(int i, String s) {
                CommonUtil.returnError(result,i,s);
            }
        });
    }
    public void deleteConversation(MethodCall methodCall, final MethodChannel.Result result) {
        // session id
        String conversationID = CommonUtil.getParam(methodCall, result, "conversationID");
        V2TIMManager.getConversationManager().deleteConversation(conversationID, new V2TIMCallback() {
            @Override
            public void onError(int i, String s) {
                CommonUtil.returnError(result,i,s);
            }

            @Override
            public void onSuccess() {
                CommonUtil.returnSuccess(result,null);
            }
        });

    }
    public void setConversationDraft(MethodCall methodCall, final MethodChannel.Result result) {
        // session id
        String conversationID = CommonUtil.getParam(methodCall, result, "conversationID");
        String draftText = CommonUtil.getParam(methodCall, result, "draftText");
        if(draftText==""){
            draftText = null;
        }
        V2TIMManager.getConversationManager().setConversationDraft(conversationID,draftText, new V2TIMCallback() {
            @Override
            public void onError(int i, String s) {
                CommonUtil.returnError(result,i,s);
            }

            @Override
            public void onSuccess() {
                CommonUtil.returnSuccess(result,null);
            }
        });

    }
    public void setConversationCustomData(MethodCall methodCall, final MethodChannel.Result result) {
        List<String> conversationIDList = methodCall.argument("conversationIDList");
        String customData = CommonUtil.getParam(methodCall, result, "customData");
        V2TIMManager.getConversationManager().setConversationCustomData(conversationIDList, customData, new V2TIMValueCallback<List<V2TIMConversationOperationResult>>() {
            @Override
            public void onSuccess(List<V2TIMConversationOperationResult> v2TIMConversationOperationResults) {
                LinkedList<HashMap<String,Object>> list = new LinkedList<>();
                for(int i = 0;i<v2TIMConversationOperationResults.size();i++){
                    list.add(CommonUtil.convertV2TIMConversationOperationResultToMap(v2TIMConversationOperationResults.get(i)));
                }
                CommonUtil.returnSuccess(result,list);
            }

            @Override
            public void onError(int i, String s) {
                CommonUtil.returnError(result,i,s);
            }
        });
    }
    public void getConversationListByFilter(MethodCall methodCall, final MethodChannel.Result result) {
        final Map<String,Object> filter = CommonUtil.getParam(methodCall,result,"filter");
        final V2TIMConversationListFilter filterNative = new V2TIMConversationListFilter();
        if(filter.get("conversationType")!=null){
            filterNative.setConversationType((Integer) filter.get("conversationType"));
        }
        if(filter.get("nextSeq")!=null){
            filterNative.setNextSeq( Long.parseLong(filter.get("nextSeq").toString()) );
        }
        if(filter.get("count")!=null){
            filterNative.setCount((Integer) filter.get("count"));
        }
        if(filter.get("markType")!=null){
            filterNative.setMarkType(0x1l << (int)filter.get("markType"));
        }if(filter.get("groupName")!=null){
            filterNative.setGroupName((String) filter.get("groupName"));
        }
        V2TIMManager.getConversationManager().getConversationListByFilter(filterNative, new V2TIMValueCallback<V2TIMConversationResult>() {
            @Override
            public void onSuccess(V2TIMConversationResult v2TIMConversationResult) {
                CommonUtil.returnSuccess(result,CommonUtil.convertV2TIMConversationResultToMap(v2TIMConversationResult));
            }

            @Override
            public void onError(int i, String s) {
                CommonUtil.returnError(result,i,s);
            }
        });

    }
    public void markConversation(MethodCall methodCall, final MethodChannel.Result result) {
        List<String> conversationIDList = methodCall.argument("conversationIDList");
        Boolean enableMark = CommonUtil.getParam(methodCall, result, "enableMark");
        int markType = CommonUtil.getParam(methodCall, result, "markType");
        long longMarkType = 0x1l << markType;
        V2TIMManager.getConversationManager().markConversation(conversationIDList, longMarkType, enableMark, new V2TIMValueCallback<List<V2TIMConversationOperationResult>>() {
            @Override
            public void onSuccess(List<V2TIMConversationOperationResult> v2TIMConversationOperationResults) {
                LinkedList<HashMap<String,Object>> list = new LinkedList<HashMap<String, Object>>();
                for(int i = 0;i<v2TIMConversationOperationResults.size();i++){
                    list.add(CommonUtil.convertV2TIMConversationOperationResultToMap(v2TIMConversationOperationResults.get(i)));
                }
                CommonUtil.returnSuccess(result,list);
            }

            @Override
            public void onError(int i, String s) {
                CommonUtil.returnError(result,i,s);
            }
        });

    }
    public void createConversationGroup(MethodCall methodCall, final MethodChannel.Result result) {
        List<String> conversationIDList = methodCall.argument("conversationIDList");
        String groupName = CommonUtil.getParam(methodCall, result, "groupName");
        V2TIMManager.getConversationManager().createConversationGroup(groupName, conversationIDList, new V2TIMValueCallback<List<V2TIMConversationOperationResult>>() {
            @Override
            public void onSuccess(List<V2TIMConversationOperationResult> v2TIMConversationOperationResults) {
                LinkedList<HashMap<String,Object>> list = new LinkedList<HashMap<String, Object>>();
                for(int i = 0;i<v2TIMConversationOperationResults.size();i++){
                    list.add(CommonUtil.convertV2TIMConversationOperationResultToMap(v2TIMConversationOperationResults.get(i)));
                }
                CommonUtil.returnSuccess(result,list);
            }

            @Override
            public void onError(int i, String s) {
                CommonUtil.returnError(result,i,s);
            }
        });
    }
    public void getConversationGroupList(MethodCall methodCall, final MethodChannel.Result result) {
        V2TIMManager.getConversationManager().getConversationGroupList(new V2TIMValueCallback<List<String>>() {
            @Override
            public void onSuccess(List<String> strings) {
                List<String>  list = new LinkedList();
                for(int i = 0;i < strings.size();i++){
                    list.add(strings.get(i));
                }
                System.out.println(list);
                CommonUtil.returnSuccess(result,list);
            }

            @Override
            public void onError(int i, String s) {
                CommonUtil.returnError(result,i,s);
            }
        });
    }
    public void deleteConversationGroup(MethodCall methodCall, final MethodChannel.Result result) {
        String groupName = CommonUtil.getParam(methodCall, result, "groupName");
        V2TIMManager.getConversationManager().deleteConversationGroup(groupName, new V2TIMCallback() {
            @Override
            public void onSuccess() {
                CommonUtil.returnSuccess(result,null);
            }

            @Override
            public void onError(int i, String s) {
                CommonUtil.returnError(result,i,s);
            }
        });
    }
    public void renameConversationGroup(MethodCall methodCall, final MethodChannel.Result result) {
        String oldName = CommonUtil.getParam(methodCall, result, "oldName");
        String newName = CommonUtil.getParam(methodCall, result, "newName");
        V2TIMManager.getConversationManager().renameConversationGroup(oldName, newName, new V2TIMCallback() {
            @Override
            public void onSuccess() {
                CommonUtil.returnSuccess(result,null);
            }

            @Override
            public void onError(int i, String s) {
                CommonUtil.returnError(result,i,s);
            }
        });

    }
    public void addConversationsToGroup(MethodCall methodCall, final MethodChannel.Result result) {
        List<String> conversationIDList = methodCall.argument("conversationIDList");
        String groupName = CommonUtil.getParam(methodCall, result, "groupName");
        V2TIMManager.getConversationManager().addConversationsToGroup(groupName,conversationIDList , new V2TIMValueCallback<List<V2TIMConversationOperationResult>>() {
            @Override
            public void onSuccess(List<V2TIMConversationOperationResult> v2TIMConversationOperationResults) {
                LinkedList<HashMap<String,Object>> list = new LinkedList<HashMap<String, Object>>();
                for(int i = 0;i<v2TIMConversationOperationResults.size();i++){
                    list.add(CommonUtil.convertV2TIMConversationOperationResultToMap(v2TIMConversationOperationResults.get(i)));
                }
                CommonUtil.returnSuccess(result,list);
            }

            @Override
            public void onError(int i, String s) {
                CommonUtil.returnError(result,i,s);
            }
        });
    }
    public void deleteConversationsFromGroup(MethodCall methodCall, final MethodChannel.Result result) {
        List<String> conversationIDList = methodCall.argument("conversationIDList");
        String groupName = CommonUtil.getParam(methodCall, result, "groupName");
        V2TIMManager.getConversationManager().deleteConversationsFromGroup( groupName,conversationIDList , new V2TIMValueCallback<List<V2TIMConversationOperationResult>>() {
            @Override
            public void onSuccess(List<V2TIMConversationOperationResult> v2TIMConversationOperationResults) {
                LinkedList<HashMap<String,Object>> list = new LinkedList<HashMap<String, Object>>();
                for(int i = 0;i<v2TIMConversationOperationResults.size();i++){
                    list.add(CommonUtil.convertV2TIMConversationOperationResultToMap(v2TIMConversationOperationResults.get(i)));
                }
                CommonUtil.returnSuccess(result,list);
            }

            @Override
            public void onError(int i, String s) {
                CommonUtil.returnError(result,i,s);
            }
        });
    }



}

File: Flutter/IMSDK/im-flutter-plugin/tencent_im_sdk_plugin/ios/Classes/manager/ConversationManager.swift
//
//  ConversationManager.swift
//  tencent_im_sdk_plugin
//
//  Created by 林智 on 2020/12/24.
//

import Foundation
import ImSDK_Plus

class ConversationManager {
	var channel: FlutterMethodChannel
	
	init(channel: FlutterMethodChannel) {
		self.channel = channel
	}

	//MARK: -Conversation
	public func getConversationList(call: FlutterMethodCall, result: @escaping FlutterResult) {
		if let nextSeq = CommonUtils.getParam(call: call, result: result, param: "nextSeq") as? String,
		   let count = CommonUtils.getParam(call: call, result: result, param: "count") as? Int32 {
            V2TIMManager.sharedInstance().getConversationList(UInt64(nextSeq) ?? 0 , count: count, succ: {
				conversations, nextSeq, finished in
				
                let dict = V2ConversationResultEntity.init(conversations: conversations ?? [], nextSeq: String(nextSeq), finished: finished).getDict();
				CommonUtils.resultSuccess(call: call, result: result, data: dict);
			}, fail: TencentImUtils.returnErrorClosures(call: call, result: result))
		}
	}
	
	public func getConversationListByConversaionIds(call: FlutterMethodCall, result: @escaping FlutterResult) {
		if let conversationIDList = CommonUtils.getParam(call: call, result: result, param: "conversationIDList") as? [String] {
		   V2TIMManager.sharedInstance().getConversationList(conversationIDList, succ: {
               if($0 != nil){
                   let dict = $0!.map { V2ConversationEntity.getDict(info: $0) };
                   CommonUtils.resultSuccess(call: call, result: result, data: dict);
               }
				
			}, fail: TencentImUtils.returnErrorClosures(call: call, result: result))
		}
	}
	
	public func getConversation(call: FlutterMethodCall, result: @escaping FlutterResult) {
		if let conversationID = CommonUtils.getParam(call: call, result: result, param: "conversationID") as? String {
			V2TIMManager.sharedInstance().getConversation(conversationID, succ: {
				conversation in
				
				CommonUtils.resultSuccess(call: call, result: result, data: V2ConversationEntity.getDict(info: conversation!));
			}, fail: TencentImUtils.returnErrorClosures(call: call, result: result))
		}
	}
	
	public func deleteConversation(call: FlutterMethodCall, result: @escaping FlutterResult) {
		let conversationID = CommonUtils.getParam(call: call, result: result, param: "conversationID") as? String;
		
		V2TIMManager.sharedInstance()?.deleteConversation(conversationID, succ: {
			() -> Void in
			
			CommonUtils.resultSuccess(call: call, result: result)
		}, fail: TencentImUtils.returnErrorClosures(call: call, result: result));
	}
	
	public func setConversationDraft(call: FlutterMethodCall, result: @escaping FlutterResult) {
		let conversationID = CommonUtils.getParam(call: call, result: result, param: "conversationID") as? String;
		let draftText = CommonUtils.getParam(call: call, result: result, param: "draftText") as? String;
		
		V2TIMManager.sharedInstance()?.setConversationDraft(conversationID, draftText: draftText, succ: {
			() -> Void in
			
			CommonUtils.resultSuccess(call: call, result: result)
		}, fail: TencentImUtils.returnErrorClosures(call: call, result: result))
	}
	
	public func pinConversation(call: FlutterMethodCall, result: @escaping FlutterResult) {
		let conversationID = CommonUtils.getParam(call: call, result: result, param: "conversationID") as? String;
		let isPinned = CommonUtils.getParam(call: call, result: result, param: "isPinned") as? Bool;
		
		V2TIMManager.sharedInstance()?.pinConversation(conversationID, isPinned: isPinned ?? false, succ: {
			() -> Void in
			CommonUtils.resultSuccess(call: call, result: result, data: "ok")
		}, fail: TencentImUtils.returnErrorClosures(call: call, result: result))
	}
	
	public func getTotalUnreadMessageCount(call: FlutterMethodCall, result: @escaping FlutterResult) {
		V2TIMManager.sharedInstance()?.getTotalUnreadMessageCount({
			totalCount -> Void in
			CommonUtils.resultSuccess(call: call, result: result, data: totalCount)
		}, fail: TencentImUtils.returnErrorClosures(call: call, result: result))
	}
    public func setConversationCustomData(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let customData = CommonUtils.getParam(call: call, result: result, param: "customData") as? String;
        let conversationIDList:[String] = CommonUtils.getParam(call: call, result: result, param: "conversationIDList") as! [String];
        V2TIMManager.sharedInstance().setConversationCustomData(conversationIDList, customData: customData?.data(using: String.Encoding.utf8)) { result_list in
            var dd : [[String:Any]] = [];
            result_list?.forEach({ item in
                dd.append([
                    "conversationID":item.conversationID ?? "",
                    "resultCode":item.resultCode ,
                    "resultInfo":item.resultInfo ?? "",
                ])
            })
            CommonUtils.resultSuccess(call: call, result: result, data: dd)
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    public func getConversationListByFilter(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let filter = CommonUtils.getParam(call: call, result: result, param: "filter") as? [String:Any] ?? [:];
        let filter_native = V2TIMConversationListFilter.init();
      
        if(filter["groupName"] as? String != nil){
            filter_native.groupName = (filter["groupName"] as? String)
        }
        
        if(filter["conversationType"]  as? Int != nil){
            filter_native.type = V2TIMConversationType.init(rawValue: Int(filter["conversationType"] as! Int)) ?? V2TIMConversationType.C2C
        }

        if(filter["nextSeq"]  as? UInt64 != nil){
            filter_native.nextSeq = (filter["nextSeq"] as! UInt64)
        }
        if(filter["count"]  as? UInt32 != nil){
            filter_native.count = (filter["count"] as! UInt32)
        }
        if(filter["markType"]  as? UInt32 != nil){
            let u64: UInt64 = 1;
            let ab = u64 << (filter["markType"] as! UInt32)
            filter_native.markType = UInt(ab);
        }
        V2TIMManager.sharedInstance().getConversationList(by: filter_native) { conversations, nextSeq, finished in
            let dict = V2ConversationResultEntity.init(conversations: conversations ?? [], nextSeq: String(nextSeq), finished: finished).getDict();
            CommonUtils.resultSuccess(call: call, result: result, data: dict);
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    public func markConversation(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let conversationIDList:[String] = CommonUtils.getParam(call: call, result: result, param: "conversationIDList") as! [String];
        let enableMark = (CommonUtils.getParam(call: call, result: result, param: "enableMark") as? Bool)!;
        let markType = (CommonUtils.getParam(call: call, result: result, param: "markType") as? Int32)!;
        let u64: UInt64 = 1;
        V2TIMManager.sharedInstance().markConversation(conversationIDList, markType: (u64 << markType) as NSNumber?, enableMark: enableMark) { result_list in
            var dd : [[String:Any]] = [];
            result_list?.forEach({ item in
                dd.append([
                    "conversationID":item.conversationID ?? "",
                    "resultCode":item.resultCode ,
                    "resultInfo":item.resultInfo ?? "",
                ])
            })
            CommonUtils.resultSuccess(call: call, result: result, data: dd)
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }


        
    }
    public func createConversationGroup(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let conversationIDList:[String] = CommonUtils.getParam(call: call, result: result, param: "conversationIDList") as! [String];
        let groupName = (CommonUtils.getParam(call: call, result: result, param: "groupName") as? String)!;
        V2TIMManager.sharedInstance().createConversationGroup(groupName, conversationIDList: conversationIDList) { result_list in
            var dd : [[String:Any]] = [];
            result_list?.forEach({ item in
                dd.append([
                    "conversationID":item.conversationID ?? "",
                    "resultCode":item.resultCode ,
                    "resultInfo":item.resultInfo ?? "",
                ])
            })
            CommonUtils.resultSuccess(call: call, result: result, data: dd)
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    public func getConversationGroupList(call: FlutterMethodCall, result: @escaping FlutterResult) {
        V2TIMManager.sharedInstance().getConversationGroupList { ids in
            CommonUtils.resultSuccess(call: call, result: result, data: ids ?? [])
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    public func deleteConversationGroup(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let groupName = (CommonUtils.getParam(call: call, result: result, param: "groupName") as? String)!;
        V2TIMManager.sharedInstance().deleteConversationGroup(groupName) {
            CommonUtils.resultSuccess(call: call, result: result)
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    public func renameConversationGroup(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let oldName = (CommonUtils.getParam(call: call, result: result, param: "oldName") as? String)!;
        let newName = (CommonUtils.getParam(call: call, result: result, param: "newName") as? String)!;
        V2TIMManager.sharedInstance().renameConversationGroup(oldName, newName: newName) {
            CommonUtils.resultSuccess(call: call, result: result)
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    public func addConversationsToGroup(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let conversationIDList:[String] = CommonUtils.getParam(call: call, result: result, param: "conversationIDList") as! [String];
        let groupName = (CommonUtils.getParam(call: call, result: result, param: "groupName") as? String)!;
        V2TIMManager.sharedInstance().addConversations(toGroup: groupName, conversationIDList: conversationIDList) { result_list in
            var dd : [[String:Any]] = [];
            result_list?.forEach({ item in
                dd.append([
                    "conversationID":item.conversationID ?? "",
                    "resultCode":item.resultCode ,
                    "resultInfo":item.resultInfo ?? "",
                ])
            })
            CommonUtils.resultSuccess(call: call, result: result, data: dd)
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

        
    }
    public func deleteConversationsFromGroup(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let conversationIDList:[String] = CommonUtils.getParam(call: call, result: result, param: "conversationIDList") as! [String];
        let groupName = (CommonUtils.getParam(call: call, result: result, param: "groupName") as? String)!;
        V2TIMManager.sharedInstance().deleteConversations(fromGroup: groupName, conversationIDList: conversationIDList) { result_list in
            var dd : [[String:Any]] = [];
            result_list?.forEach({ item in
                dd.append([
                    "conversationID":item.conversationID ?? "",
                    "resultCode":item.resultCode ,
                    "resultInfo":item.resultInfo ?? "",
                ])
            })
            CommonUtils.resultSuccess(call: call, result: result, data: dd)
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }
        
    }
}
